File: JBGE_Unity/Assets/JBGE_Unity/GameEngine/GameEngine.cs
using System;
using System.Collections.Generic;
using System.IO;
using UnityEngine;
using UnityEngine.AddressableAssets;
using UnityEngine.InputSystem;
using UnityEngine.ResourceManagement.ResourceLocations;
using UnityEngine.UIElements;
using System.Collections;
using UnityEngine.SceneManagement;

namespace JBGE {
	public class GameEngine {
    // Main GameObject
    public GameObject MainGameObject;

    // Main Perspective Camera
    public Camera MainCamera;

    // Manages the Ui components
    public GameObject UIManager;
    // Manages the non-UI components
    public GameObject SceneManager;

    // Flag to determine if DEBUG info should be shown on screen or not
    public bool IsShowDebugInfo { get; set; } = false;
    // Defines the target frame rate (defaults to 60)
    public int TargetFrameRate { get; set; } = 60;
    // Stores the average frame rate
    public int AverageFrameRate { get; set; }
    // Stores the frame count 0 to 60
    public int FrameCount { get; set; }
    // Amount of frame to wait until executing the next script
    public int WaitFrameCount { get; set; }
    // Determines if we are in wait state or not
    public bool IsWaiting { get; set; }

    // The Pixel Per Unit scale to lift up UI scaling
    public float PPUScaleUpUI = 4f;
		// The Pixel Per Unit scale to lift up 2D Map scaling (Actually it's the ortho size so not affecting the actual object scales)
		public float PPUScaleUpWorld = 5f;
    // Screen width in pixels
    public float ScreenWidth = 0f;
    // Screen height in pixels
    public float ScreenHeight = 0f;
    // If in perspective mode, the horizontal Field Of View Angle (in degrees)
    public float FOV = 60f;

    // The one and only Base UI Layer that always sits in front of the UICamera
    public UIComponent UIBaseLayer;
    // UI layer that manages actors (or sprites)
    public UIComponent UIBackgroundLayer;

		// Stores all the UILayer(s) that is placed on the UIBaseLayer UICanvas
		public Dictionary<int, UIComponent> UILayers = new Dictionary<int, UIComponent>();

    // User gamepad device inputs
		public IDGamePad IDGamePad = new IDGamePad();
    // User mouse device inputs
    public IDMouse IDMouse = new IDMouse();
    // User keyboard device inputs
    public IDKeyboard IDKeyboard = new IDKeyboard();
    // Manages all user inputs
    public IDUserInput UserInput;

    public GameEngine(GameObject gameObject, float screenWidth, float screenHeight) {
      MainGameObject = gameObject;

			// Setup main camera
			MainCamera = GameObject.Find("Main Camera").GetComponent<Camera>();
			MainCamera.orthographic = false;
      MainCamera.farClipPlane = 1000;
      MainCamera.nearClipPlane = 0.01f;
			MainCamera.usePhysicalProperties = false;
			MainCamera.fieldOfView = FOV;

      SceneManager = new GameObject("SceneManager");
      SceneManager.transform.SetParent(MainGameObject.transform);
      UIManager = new GameObject("UIManager");
      UIManager.transform.SetParent(MainGameObject.transform);

			// We need to call this in order to check for Addressable asset existence
			Addressables.InitializeAsync();

      UpdateScreenSize(screenWidth, screenHeight);

			// Initialize UI
			InitializeUI();
		}

    private void InitializeUI() {
			// Create user input object
			UserInput = new IDUserInput(this);

			// Create the one and utmost base layer that attaches to the UICamera so we can place UILayer(s) and its child UI components accordingly
			UIBaseLayer = new UIComponent(this, "UIBaseLayer", null, true);
      // Add UIBaseLayer to UIManager as root layer
      UIBaseLayer.ThisObject.transform.SetParent(UIManager.transform);

      // Test
      /*
      UIBaseLayer.TransformAll(
            new Vector3(0.5f, 0.5f, 0.5f),
            new Vector3(0.0f, 0.0f, 0.0f),
            new Vector3(0.5f, 0.5f, 1.0f),
            new Vector3(0, 0, 45),
            new Vector2(0.0f, 0.0f),
            new Vector2(0.0f, 0.0f),
            new Vector2(1.0f, 1.0f));*/

			// Create UI Layers
			//int UIBackgroundLayerID = CreateUILayer("UIBackgroundLayer");
      //UIBackgroundLayer = UILayers[UIBackgroundLayerID];
		}

    // Called when screen size changes
    public void UpdateScreenSize(float width, float height) {
        if(height <= 0) return;
        ScreenWidth = width;
        ScreenHeight = height;
        Debug.Log($"[GameEngine] Screen Size Changed: {width} x {height} --> aspect: {width / height}");
    }

    public void Update() {
      UserInput.Update();
    }

		/// <summary>
		/// Check if asset exists in addressable or not
		/// IMPORTANT: Before this method can work, we need to call: Addressables.InitializeAsync();
		/// at the very start of app initialization in order to check for Addressable asset existence
		/// </summary>
		/// <param name="key">The Addressable key path to check for existence</param>
		public static bool IsAddressableAssetExists(object key) {
      if(Application.isPlaying) {
        foreach(var l in Addressables.ResourceLocators) {
          IList<IResourceLocation> locs;
          if(l.Locate(key, null, out locs)) return true;
        }
        return false;
      }
      return false;
    }

    /// <summary>Load asset from addressables</summary>
    /// <typeparam name="T">The asset type to load</typeparam>
    /// <param name="address">The address path of the asset</param>
    /// <returns>Loaded asset as object</returns>
    public object LoadAddressableAsset<T>(string address) {
      // If asset does not exist in the addressables, return null
      if(!IsAddressableAssetExists(address)) return null;
      // For Unity, if asset is PNG, we want to use the ".bytes" extension (unless, it converts to Texture2D which we don't want to)
      if(Path.GetExtension(address) == ".png") {
        address = Path.ChangeExtension(address, ".bytes");
      }
      try {
        var handle = Addressables.LoadAssetAsync<T>(address);
        var asset = handle.WaitForCompletion();
        Addressables.Release(handle);
        return asset;
      } catch {
        Debug.Log("Failed to load asset from Addressables: " + address);
			}
      return null;
    }


    /// <summary>Creates a new Layer under the base layer</summary>
    /// <returns>ID is generated that can be used to identify the newly created object (if creation fails, returns -1)</returns>
    public int CreateUILayer(string layerName = "Layer") {
      UIComponent layer = new UIComponent(this, layerName, UIBaseLayer);
			// Sort order for each custom UI layer is incremented by 1000, where the UI BaseLayer is 0
			layer.SortOrder = (UILayers.Count + 1) * 1000;
      // Check if same ID already exists in our list
      while(UILayers.ContainsKey(layer.ID)) {
        // Re-generate a random ID for this object
        layer.ID = new System.Random().Next();
      }
      UILayers.Add(layer.ID, layer);
      return layer.ID;
    }

    /// <summary>Destroys the UILayer from our list</summary>
    /// <param name="id">The handle ID of the object to be deleted</param>
    public void DestroyUILayer(int id) {
      if(!UILayers.ContainsKey(id)) return;
      UILayers[id].Destroy();
      UILayers[id] = null;
      UILayers.Remove(id);
    }
	}
}

File: JBGE_RealityKit/JBGE_RealityKit/GameEngine/GameEngine.swift
//
//  GameEngine.swift
//  JBGE_RealityKit
//
//  Created by Tomohiro Kadono on 2026/01/03.
//

import RealityKit

open class GameEngine {
    // Main GameObject
    public var MainGameObject: Entity
    
    // Main Perspective Camera
    public var MainCamera: Entity

    // Manages the UI components
    public var UIManager: Entity? = nil
    // Manages the non-UI components
    public var SceneManager: Entity? = nil
    
    // Flag to determine if DEBUG info should be shown on screen or not
    public var IsShowDebugInfo: Bool = false
    // Defines the target frame rate (defaults to 60)
    public var TargetFrameRate: Int = 60
    // Stores the average frame rate
    public var AverageFrameRate: Int = 0
    // Stores the frame count 0 to 60
    public var FrameCount: Int = 0
    // Amount of frame to wait until executing the next script
    public var WaitFrameCount: Int = 0
    // Determines if we are in wait state or not
    public var IsWaiting: Bool = false

    // The Pixel Per Unit scale to lift up UI scaling
    public var PPUScaleUpUI: Float = 4.0
    // The Pixel Per Unit scale to lift up 2D Map scaling (Actually it's the ortho size so not affecting the actual object scales)
    public var PPUScaleUpWorld: Float = 5.0
    // Screen width in pixels
    public var ScreenWidth: Float = 0
    // Screen height in pixels
    public var ScreenHeight: Float = 0
    // If in perspective mode, the horizontal Field Of View Angle (in degrees)
    public var FOV: Float = 60.0
    
    // The one and only Base UI Layer that always sits in front of the UICamera
    public var UIBaseLayer: UIComponent? = nil
    // UI layer that manages actors (or sprites)
    public var UIBackgroundLayer: UIComponent? = nil
    
    // Stores all the UILayer(s) that is placed on the UIBaseLayer UICanvas
    public var UILayers: [Int: UIComponent] = [:]

    // User gamepad device inputs
    public var IDGamePad: IDGamePad = JBGE_RealityKit.IDGamePad()
    // User mouse device inputs
    public var IDMouse: IDMouse = JBGE_RealityKit.IDMouse()
    // User keyboard device inputs
    public var IDKeyboard: IDKeyboard = JBGE_RealityKit.IDKeyboard()
    // Manages all user inputs
    public lazy var UserInput: IDUserInput = IDUserInput(self)
    
    public init(_ gameObject: Entity, _ screenWidth: Float, _ screenHeight: Float) {
        MainGameObject = gameObject
        
        // Setup main camera
        var cameraEntity = Entity()
        cameraEntity.name = "MainCamera"
        var camera = PerspectiveCameraComponent()
        camera.fieldOfViewInDegrees = FOV
        camera.near = 0.01
        camera.far  = 1000.0
        cameraEntity.components.set(camera)
        cameraEntity.transform.translation = SIMD3<Float>(0, 0, 1.5)
        MainCamera = cameraEntity
        MainGameObject.addChild(MainCamera)
        
        SceneManager = Entity()
        SceneManager?.name = "SceneManager"
        MainGameObject.addChild(SceneManager ?? Entity())
        UIManager = Entity()
        UIManager?.name = "UIManager"
        MainGameObject.addChild(UIManager ?? Entity())
        
        // TEST
        if let cube = MainGameObject.findEntity(named: "CubeTest") {
            cube.removeFromParent()
            SceneManager?.addChild(cube)
        }
        
        UpdateScreenSize(screenWidth, screenHeight)
        
        // Initialize UI
        InitializeUI()
    }

    private func InitializeUI() {
        // Create user input object
        UserInput = IDUserInput(self);
        
        // Create the one and utmost base layer
        UIBaseLayer = UIComponent(self, "UIBaseLayer", nil, true)
        // Add UIBaseLayer to UIManager as root layer
        UIManager?.addChild(UIBaseLayer?.ThisObject ?? Entity())
        // Test 0:
        /*UIBaseLayer?.TransformAll(
            SIMD3<Float>(0.5, 0.5, 1),
            SIMD3<Float>(0.0, 0.0, 0.0),
            SIMD3<Float>(0.5, 0.5, 1.0),
            SIMD3<Float>(0, 0, 45),
            SIMD2<Float>(0.0, 0.0),
            SIMD2<Float>(0.0, 0.0),
            SIMD2<Float>(1.0, 1.0))
        // Test 1:
        UIBaseLayer?.TransformAll(
            SIMD3<Float>(1, 1, 1),
            SIMD3<Float>(0, 0, 0),
            SIMD3<Float>(1, 1, 1),
            SIMD3<Float>(0, 0, 0),
            SIMD2<Float>(0, 0),
            SIMD2<Float>(0.5, 0.5),
            SIMD2<Float>(0.5, 0.5)
        );
        // Test 2:
        UIBaseLayer?.TransformAll(
            SIMD3<Float>(0.5, 0.5, 1),
            SIMD3<Float>(0.5, 0.5, 0),
            SIMD3<Float>(1, 1, 1),
            SIMD3<Float>(0, 0, 0),
            SIMD2<Float>(0.5, 0.5),
            SIMD2<Float>(0.5, 0.5),
            SIMD2<Float>(0.5, 0.5)
        );
        // Test 3:
        UIBaseLayer?.TransformAll(
            SIMD3<Float>(0.5, 0.5, 1),
            SIMD3<Float>(1, 1, 0),
            SIMD3<Float>(1, 1, 1),
            SIMD3<Float>(0, 0, 0),
            SIMD2<Float>(1, 1),
            SIMD2<Float>(0.5, 0.5),
            SIMD2<Float>(0.5, 0.5)
        );
        // Test 4:
        UIBaseLayer?.TransformAll(
            SIMD3<Float>(1, 1, 1),
            SIMD3<Float>(0, 0, 0),
            SIMD3<Float>(0.5, 0.5, 1),
            SIMD3<Float>(0, 0, 0),
            SIMD2<Float>(0, 0),
            SIMD2<Float>(0, 0),
            SIMD2<Float>(0.5, 0.5)
        );
        // Test 5:
        UIBaseLayer?.TransformAll(
            SIMD3<Float>(0.6, 0.6, 1),
            SIMD3<Float>(0.5, 0.5, 0),
            SIMD3<Float>(1, 1, 1),
            SIMD3<Float>(0, 0, 30),
            SIMD2<Float>(0.5, 0.5),
            SIMD2<Float>(0.5, 0.5),
            SIMD2<Float>(0.5, 0.5)
        );
        // Test 6:
        UIBaseLayer?.TransformAll(
            SIMD3<Float>(0.5, 0.5, 1),
            SIMD3<Float>(0, 0, 0),
            SIMD3<Float>(0.5, 0.5, 1),
            SIMD3<Float>(0, 0, 45),
            SIMD2<Float>(0, 0),
            SIMD2<Float>(0, 0),
            SIMD2<Float>(1, 1)
        );
        // Test 7:
        UIBaseLayer?.TransformAll(
            SIMD3<Float>(0.75, 0.3, 1),
            SIMD3<Float>(0.4, 0.1, 0),
            SIMD3<Float>(0.5, 0.5, 1),
            SIMD3<Float>(0, 0, 0),
            SIMD2<Float>(0.25, 0.25),
            SIMD2<Float>(0.8, 0.8),
            SIMD2<Float>(0.5, 0.5)
        );*/
        // Test 8:
        UIBaseLayer?.TransformAll(
            SIMD3<Float>(0.75, 0.3, 1),
            SIMD3<Float>(0.4, 0.1, 0),
            SIMD3<Float>(0.5, 0.5, 1),
            SIMD3<Float>(20, 10, 80),
            SIMD2<Float>(0.25, 0.25),
            SIMD2<Float>(0.8, 0.8),
            SIMD2<Float>(0.16, 0.23)
        );
        
        // Create UI Layers
        //let UIBackgroundLayerID = CreateUILayer("UIBackgroundLayer")
        //UIBackgroundLayer = UILayers[UIBackgroundLayerID]

        print("[GameEngine] Initialize UI Completed.")
        print("===== ENTITY HIERARCHY DUMP =====")
        dumpEntityTree(MainGameObject)
        print("================================")
    }
    
    // Called when screen size changes
    public func UpdateScreenSize(_ width: Float, _ height: Float) {
        guard height > 0 else { return }
        ScreenWidth = width
        ScreenHeight = height
        print("[GameEngine] Screen Size Changed: \(width) x \(height) --> aspect: \(width / height)")
    }

    /// <summary>Creates a new Layer under the base layer</summary>
    /// <returns>ID is generated that can be used to identify the newly created object (if creation fails, returns -1)</returns>
    @discardableResult
    public func CreateUILayer(_ layerName: String = "Layer") -> Int {
        guard let baseLayer = UIBaseLayer else { return -1 }
        let layer = UIComponent(self, layerName, baseLayer, true)
        // Sort order for each custom UI layer is incremented by 1000, where the UI BaseLayer is 0
        layer.SortOrder = (UILayers.count + 1) * 1000
        // Ensure unique ID
        while UILayers[layer.ID] != nil {
            layer.ID = Int.random(in: Int.min...Int.max)
        }
        UILayers[layer.ID] = layer
        return layer.ID
    }

    /// <summary>Destroys the UILayer from our list</summary>
    /// <param name="id">The handle ID of the object to be deleted</param>
    public func DestroyUILayer(_ id: Int) {
        guard let layer = UILayers[id] else { return }
        layer.Destroy()
        UILayers.removeValue(forKey: id)
    }
    
    private var time: Float = 0

    public func Update(_ deltaTime: Float) {
        time += deltaTime

        guard let scene = SceneManager else { return }

        let r: Float = 2.0
        let x = cos(time) * r
        let z = sin(time) * r

        // ワールドを動かす（＝カメラが周回して見える）
        scene.position = SIMD3<Float>(-x, 0, -z)
    }
    
    // Debug: Dumps the entire scene's hierarchy
    private func dumpEntityTree(
        _ entity: Entity,
        indent: String = "",
        isLast: Bool = true
    ) {
        let marker = isLast ? "└─" : "├─"
        let name = entity.name.isEmpty ? "(unnamed)" : entity.name
        let typeName = String(describing: type(of: entity))

        print("\(indent)\(marker) \(name) [\(typeName)]")

        let nextIndent = indent + (isLast ? "   " : "│  ")
        let children = entity.children

        for i in 0..<children.count {
            dumpEntityTree(
                children[i],
                indent: nextIndent,
                isLast: i == children.count - 1
            )
        }
    }
}

